lib/nostrum/api.ex

```diff
@@ -138,51 +138,108 @@ defmodule Nostrum.Api do
   end
 
   @doc ~S"""
-  Send a message to a channel.
+  Posts a message to a guild text or DM channel.
 
-  ## Parameters
-    - `channel_id` - Id of the channel to send the message to.
-    - `content` - One of string, embed, or file to send to the channel.
-    See `t:message_content/0` for more info.
-    - `tts` - Whether the message should be read over text to speech.
+  This endpoint requires the `VIEW_CHANNEL` and `SEND_MESSAGES` permissions. It
+  may situationally need the `SEND_MESSAGES_TTS` permission. It fires the
+  `t:Nostrum.Consumer.message_create/0` event.
+
+  If successful, returns `{:ok, message}`. Otherwise, returns a `t:Nostrum.Api.error/0`.
 
-  For the `channel_id` parameter, you can pass in a `Nostrum.Struct.Message`
-  struct, and it will pull the id from there.
+  ## Options
+
+    * `:content` (string) - the message contents (up to 2000 characters)
+    * `:nonce` (`t:Nostrum.Struct.Snowflake.t/0`) - a nonce that can be used for
+    optimistic message sending
+    * `:tts` (boolean) - true if this is a TTS message
+    * `:file` (`t:Path.t/0`) - the path of the file being sent
+    * `:embed` (`t:Nostrum.Struct.Embed.t/0`) - embedded rich content
+
+    At least one of the following is required: `:content`, `:file`, `:embed`.
+
+  ## Examples
 
-  ## Example
   ```Elixir
-  Nostrum.Api.create_message(1111111111111, [content: "my os rules", file: ~S"C:\i\use\windows"])
+  Nostrum.Api.create_message(43189401384091, content: "hello world!")
+
+  import Nostrum.Struct.Embed
+  embed =
+    %Nostrum.Struct.Embed{}
+    |> put_title("embed")
+    |> put_description("new desc")
+  Nostrum.Api.create_message(43189401384091, embed: embed)
+
+  Nostrum.Api.create_message(43189401384091, file: "/path/to/file.txt")
+
+  Nostrum.Api.create_message(43189401384091, content: "hello world!", embed: embed, file: "/path/to/file.txt")
   ```
   """
-  @spec create_message(Message.t(), message_content, boolean) :: error | {:ok, Message.t()}
-  @spec create_message(Channel.id(), message_content, boolean) :: error | {:ok, Message.t()}
-  def create_message(channel_id, content, tts \\ false)
+  @spec create_message(Channel.id(), options) :: error | {:ok, Message.t()}
+  def create_message(channel_id, options)
 
-  def create_message(%Message{channel_id: id}, content, tts) when is_binary(content),
-    do: create_message(id, content, tts)
+  def create_message(channel_id, options) when is_list(options),
+    do: create_message(channel_id, Map.new(options))
 
-  # Sending regular messages
-  def create_message(channel_id, content, tts) when is_binary(content),
-    do: do_create_message(channel_id, %{content: content, tts: tts})
+  def create_message(channel_id, options) when is_snowflake(channel_id) do
+    case options do
+      %{file: _} -> create_message_with_multipart(channel_id, options)
+      %{} -> create_message_with_json(channel_id, options)
```

Comment:
Do you think it would be better to guard against `options` being a list or map? If the user passes in something garbage here for `options` I think it'd be better to say there's no matching function, as opposed to no match for the case which the user doesn't really need to worry about. We can just make this second statement match on `_`. If they send in junk within the enumerable they'll get an API error which I think is fine. Let me know your thoughts.